### PR TITLE
MNT: Use new RTD search API

### LINKED ISSUE #459: Search: using deprecated API (https://readthedocs.org/api/v2/docsearch/)
Hi! I work at Read the Docs, I'm reaching to you since we are in the process of removing this undocumented API (https://readthedocs.org/api/v2/docsearch). We are replacing it with https://docs.readthedocs.io/en/stable/server-side-search.html#api 

I noticed you are using the old API at https://github.com/astropy/astropy-tutorials/blob/21968ff28eaf1c72f9d022ceb788bb2fc464022e/tutorials/themes/tutorials-theme/static/searchtools.js#L419

You can check how we use it at https://github.com/readthedocs/readthedocs.org/blob/master/readthedocs/core/static-src/core/js/doc-embed/search.js and https://github.com/readthedocs/readthedocs-sphinx-search/blob/master/sphinx_search/static/js/rtd_sphinx_search.js

### PR DESCRIPTION
**Pull request by pllim**
I started this but I am not really involved on the JS side of things here, so if someone else wanna take over, feel free. Fix #459 

Will a simple URL search and replace be sufficient, or we need to change other stuff too? 

cc @embray

### CODE REVIEW

**Comment by kelle** [comment]
Looping in @jonathansick , this sounds like something he may have thoughts on.

File: tutorials/themes/tutorials-theme/static/searchtools.js [hunk 1]
```diff
@@ -416,17 +416,17 @@ var Search = {
       if (query.includes("filterTutorials")) { // don't search docs if we are just looking for tutorials
         var docResponse = null;
         var docResult = '';
-        $.get('https://readthedocs.org/api/v2/docsearch/?q=' + window.location.search.split('=')[1] + '&project=astropy&version=stable&language=en', function(response){
+        $.get('https://docs.readthedocs.io/_/api/v2/search/?q=' + window.location.search.split('=')[1] + '&project=astropy&version=stable', function(response){
           docResponse = response;
           //limits the maximum results from the documentation to 15
           docResultsCount = (docResponse.results.hits.total>15)? 15 : docResponse.results.hits.total ;
```
Comment (stsewd) [inline tutorials/themes/tutorials-theme/static/searchtools.js:422]: This should be just `results.length`

File: tutorials/themes/tutorials-theme/static/searchtools.js [hunk 2]
```diff
@@ -416,17 +416,17 @@ var Search = {
       if (query.includes("filterTutorials")) { // don't search docs if we are just looking for tutorials
         var docResponse = null;
         var docResult = '';
-        $.get('https://readthedocs.org/api/v2/docsearch/?q=' + window.location.search.split('=')[1] + '&project=astropy&version=stable&language=en', function(response){
+        $.get('https://docs.readthedocs.io/_/api/v2/search/?q=' + window.location.search.split('=')[1] + '&project=astropy&version=stable', function(response){
```
Comment (stsewd) [inline tutorials/themes/tutorials-theme/static/searchtools.js:419]: The new endpoint should be https://docs.astropy.org/_/api/v2/search/?q=test&project=astropy&version=stable

File: tutorials/themes/tutorials-theme/static/searchtools.js [hunk 3]
```diff
@@ -416,17 +416,17 @@ var Search = {
...
           docResponse = response;
           //limits the maximum results from the documentation to 15
           docResultsCount = (docResponse.results.hits.total>15)? 15 : docResponse.results.hits.total ;
 
           for(var i = 0; i < docResultsCount; i++ ) {
-            docResult += '<li style="" class="result-card doc"> <a href="' + 
-                              docResponse.results.hits.hits[i].fields.link + '.html' +  
+            docResult += '<li style="" class="result-card doc"> <a href="' +
+                              docResponse.results.hits.hits[i].fields.link + '.html' +
```
Comment (stsewd) [inline tutorials/themes/tutorials-theme/static/searchtools.js:426]: The link now includes the extension. `results[i].domain + results[i].path`

File: tutorials/themes/tutorials-theme/static/searchtools.js [hunk 4]
```diff
@@ -416,17 +416,17 @@ var Search = {
...
                               '?highlight=' + window.location.search.split('=')[1] + '">' +
                               docResponse.results.hits.hits[i].fields.title[0] + '</a>' +
```
Comment (stsewd) [inline tutorials/themes/tutorials-theme/static/searchtools.js:428]: this should be `results[i].title`

File: tutorials/themes/tutorials-theme/static/searchtools.js [hunk 5]
```diff
@@ -416,17 +416,17 @@ var Search = {
...
-                              '<div class="context">' + 
+                              '<div class="context">' +
                               docResponse.results.hits.hits[i].highlight.content[0] +
```
Comment (stsewd) [inline tutorials/themes/tutorials-theme/static/searchtools.js:430]: This should be iterating over `results[i].blocks`

**Review by stsewd — COMMENTED** [review]
The response of the API changed, you can take a look at the new response in https://docs.astropy.org/_/api/v2/search/?q=test&project=astropy&version=stable or check the docs at https://docs.readthedocs.io/en/stable/server-side-search.html#api

**Comment by pllim** [comment]
Thanks, @stsewd ! Just want you to know that I received your much appreciated review, but don't have time to follow-up yet.

**Comment by embray** [comment]
Sorry I didn't get to work on this sooner.  I guess it's more urgent than I remembered.

**Comment by embray** [comment]
I'm implementing @stsewd 's review comments.

**Comment by pllim** [comment]
Go for it, @embray . I'll hold off. Thank you for your help!

File: tutorials/themes/tutorials-theme/static/searchtools.js (hunk 2 above)
Comment (stsewd) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: oh, I think you can use the other endpoint https://readthedocs.org/api/v2/search/?q=test&project=astropy&version=stable let me know if that one works

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: I still get

> Access to XMLHttpRequest at 'https://readthedocs.org/api/v2/search/?q=test&project=astropy&version=stable' from origin 'http://learn.astropy.org' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

:/

Comment (stsewd) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: hmm, not sure how this was working before.

https://github.com/readthedocs/readthedocs.org/blob/12598254c86e954f19c7402395d9e402f9a2aaaa/readthedocs/core/signals.py#L33-L82

Looks like you can get around by creating a custom domain with "learn.astropy.org", there is no need to add the CNAME, this is so rtd knows this is a known domain.

Comment (stsewd) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: I mean, creating the custom domain the astropy project.

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: I'm not sure if it *was* working before.  Presumably it was at some point. But even with the current version of the site using the old/undocumented search it also gets CORS errors, so it isn't working currently either.

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: @stsewd I see.  It turns out learn.astropy.org is not even built/hosted by RTD.  I assumed it was, but apparently it is using a CI service to build the site and push it to gh-pages.  Seeing that code helps, though it's still not exactly clear to me how I can add learn.astropy.org as a known domain without hosting these docs on RTD (which I'm not opposed to, but someone who was not me apparently made the decision at one point not to).

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: Oh, I found the settings on RTD for adding an additional custom domain.  I see your point now about adding it as a custom domain without actually adding a CNAME entry.  Hopefully that should work (it does not yet, but it still shows "SSL certificate status: pending_validation" on https://readthedocs.org/dashboard/astropy/domains/10700/edit/

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: It appears to be due to this, but in this case RTD wouldn't be issuing an SSL cert on behalf of the site: https://docs.readthedocs.io/en/stable/custom_domains.html#certificate-authority-authorization

Comment (embray) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: Although looking again at the code you linked, for the CORS access it doesn't seem to care about certificate status :/

Comment (pllim) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: > someone who was not me apparently made the decision at one point not to

@adrn , @eblur , or @kelle , do you remember why this decision was made? Can you please help @embray move this forward? Thanks.

Comment (stsewd) [reply tutorials/themes/tutorials-theme/static/searchtools.js:419]: > Although looking again at the code you linked, for the CORS access it doesn't seem to care about certificate status

Yeah, you only need to add the domain, the status or the CNAME don't matter here. Let me know if that fixed the CORS problem.

**Comment by adrn** [comment]
Only ~a year late, but new front-end is launching imminently, so closing this. Thanks @pllim and @stsewd, and sorry for losing track of this...

**Pull request closed without merging**